Source/nthread.cpp

```diff
@@ -59,98 +59,93 @@ void __cdecl nthread_cleanup_mutex()
 	DeleteCriticalSection(&sgMemCrit);
 }
 
-void __fastcall nthread_terminate_game(char *pszFcn)
+void __fastcall nthread_terminate_game(const char *pszFcn)
 {
-	char *v1; // esi
-	int v2; // eax
-	char *v3; // eax
+	DWORD sErr; // eax
 
-	v1 = pszFcn;
-	v2 = SErrGetLastError();
-	if ( v2 != STORM_ERROR_INVALID_PLAYER )

```

Comment:
Does this leave the generated binary identical to original?

Reply:
(it seems I quoted too few lines. I mean the inlining these two params)

Reply:
Are you implying that the generated binary was identical to the original to begin with?
My impression is that the Hex-Rays decompiler does not necessarily produce C code that compiles to what the original was, and this here looks like decompiler goop, instead of eg. something that a human wrote to match the original binary.
(no sane human being would write such code to begin with so it can't be the "original" code)

Reply:
Umm, I am only a passive reader and I am yet to investigate myself whether `nthread_terminate_game()` in "Devilution.exe" matches the original "Diablo.exe" (at least in length of the function body in bytes).  So I don't have an answer.

As I still have no tools set up, I wonder about such parameter-shadowing vars decompiled by Hex-Rays in general (since the day I found this project).  Can it be that they are always phantom / don't occupy additional CPU instructions in function? So maybe it's @galaxyhaxz (or @mewmew) to be asked first:)

I guess there are cases when human coder would like to keep such shadow copy of a param and reuse it later.  

OK, now I start realizing that this can be detected when you find such var on the right sight of an assignment.  But my question regarding additional space in the body remains open.

Reply:
The reason why the temporary variable `v1` exists here (and in many other places) is due to register allocation that the compiler does (and which the decompiler does not undo). The original code very likely did not have that copy.
The `nthread_terminate_game` function is `fastcall`, therefore it receives an argument in the `ecx` register. Then it immediately calls `SErrGetLastError` which is `stdcall` and does not preserve `ecx` - so at this point the compiler is *forced* to spill the variable (ie. copy it to another register or stack slot).
The decompiler could be smart enough to remove that copy, but maybe it wants to preserve the information that the variable was spilled to the `esi` register or whatever.

Reply:
@nomdenom Glad you are joining the project! You seem quite knowledgeable with compiler theory, which is definitely helpful for a project like this :) You should consider joining the Discord Devilution channel if you haven't already. Ping @seritools for invite.

Reply:
@sskras the hex-ray output, recompiled with a compiler doesn't really yield the exact same binary. hex-rays often adds temporary variables representing a specific state of a register or piece of memory.

With the help of IDA/Hex-Rays and using the right compiler and service pack (VC6 SP3) you can mostly recreate the code to be binary-exact or near that at least though. Very helpful for that is looking at the PSX headers (in the case of nthread there aren't any, though) and looking at older versions of the binary (especially prerelease versions like the battle.net beta) which are way less optimized, so Hex-Rays has an easier time finding out how the code was supposed to be.

Check out my changes to monster.cpp for example:  #163 The code is basically completely rewritten and now looks like code a human would write, while simultaneously often being even closer to the original binary (since the compiler does the same optimizations in the same way as with the original source of the game).

You're right that the vast majority of "shadow vars" don't influence the performance/codegen at all, since they are optimized away.

@nomdenom Absolutely correct. Most of the work cleaning up the code is guessing what the code might have looked like before compilation/optimization, then testing that and checking/comparing the generated binary with the original one.

Reply:
Thanks guys, your explanations are wonderful !